File: Ruby/tablero.rb
# encoding: utf-8
module Civitas
  require_relative "casilla.rb"
  require_relative "tipo_casilla.rb"
  require_relative "sorpresa.rb"
  require_relative "tipo_sorpresa.rb"
  
  class Tablero
  
    def initialize(casilla)
      if(casilla>=1)
        @numCasillaCarcel = casilla
      else
        @numCasillaCarcel = 1
      end
      @casillas = Array.new
      casilla_salida = Casilla.new(TipoCasilla::DESCANSO, "Salida")
      @casillas.push(casilla_salida)
      @porSalida = 0
      @tieneJuez = false
  
    end
    
    
  
    private
    
    def correcto
      ok = false
      if(@casillas.length>@numCasillaCarcel&&@tieneJuez)
        ok = true
      end
      
      return ok
  
    end
    
    def correcto_casilla(numCasilla)
      ok = false
      if(correcto()&&numCasilla>0&&numCasilla<@casillas.length)
        ok = true
      end
    end
    
    public
    
    attr_accessor :numCasillaCarcel
    
    
    def getPorSalida
      resultado = @porSalida
      if(@porSalida > 0)
        @porSalida = @porSalida -1
      end
      return resultado
    end
    
    def compruebaCarcel
      if(@casillas.length==@numCasillaCarcel)
        casilla_carcel = Casilla.new(TipoCasilla::JUEZ, @numCasillaCarcel, "Cárcel")
        @casillas.push(casilla_carcel)
      end
    end
    
    def añadeCasilla(casilla)
      compruebaCarcel
      @casillas.push(casilla)
      compruebaCarcel
    end
    
    def añadeJuez
      if(!@tieneJuez)
        casilla_juez = Casilla.new(TipoCasilla::JUEZ, @numCasillaCarcel, "Juez")
        @casillas.push(casilla_juez)
        @tieneJuez=true
      end
    end
    
    def getCasilla(numCasilla)
      if(correcto(numCasilla))
        respuesta = @casillas[numCasilla]
      else
        respuesta = nil
      end
      return respuesta
    end
    
    def nuevaPosicion(actual, tirada)
      if(!correcto)
        respuesta = -1
      else
        nuevaPosicion = actual + tirada
      end
      if(nuevaPosicion > @casillas.length)
        nuevaPosicion = nuevaPosicion % @casillas.length - 1
        @porSalida = @porSalida + 1
      end
      return nuevaPosicion
    end
    
    
    
    def tamaño
      return @casillas.length
    end
    
    def calcularTirada(origen, destino)
      tirada = destino - origen
      if(tirada < 0)
        tirada += @casillas.length
      end
      return tirada
    end
    
    
  end
end

File: civitas/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;
import GUI.*;
/**
 *
 * @author nachobc82
 */
public class Tablero {
    private int numcasillacarcel;
    private ArrayList<Casilla> casillas;
    private int porsalida;
    private boolean tienejuez;
    
    Tablero(int carcel){
        this.numcasillacarcel = carcel;
        this.casillas = new ArrayList<>();
        this.casillas.add(new Casilla("Salida"));
        this.porsalida = 0;
        this.tienejuez = false;
    }
    
    private boolean Correcto(){
        boolean resultado = true;
        if (casillas.size() < this.numcasillacarcel || !this.tienejuez)
            resultado = false;
        return resultado;
    }
    
    private boolean Correcto(int numCasilla){
        boolean resultado = this.Correcto();
        if (numCasilla > casillas.size())
            resultado = false;
        return resultado;
    }
    
    int getCarcel(){
        return numcasillacarcel;
    }
    
    int getPorSalida(){
        int resultado;
        resultado = this.porsalida;
        if(this.porsalida > 0)
            this.porsalida--;
        return resultado;
    }
    
    void AñadeCasilla(Casilla casilla){
        if (this.casillas.size() == this.numcasillacarcel)
            this.casillas.add(new Casilla("Carcel"));
        this.casillas.add(casilla);
        if (this.casillas.size() == this.numcasillacarcel)
            this.casillas.add(new Casilla("Carcel"));
    }
    
    void AñadeJuez(){
        if( !this.tienejuez){
            CasillaJuez casilla = new CasillaJuez(this.numcasillacarcel, "Juez");
            this.AñadeCasilla(casilla);
            this.tienejuez = true;
        } 
    }
    
    Casilla getCasilla(int numCasilla){
        if(this.Correcto(numCasilla)){
            return casillas.get(numCasilla);
        }
        else
            return null;
    }
    
    int NuevaPosicion(int actual, int tirada){
        int casilla_final = -1;
        if (this.Correcto()){
            casilla_final = (actual + tirada)%casillas.size();
            if(casilla_final != actual+tirada)
                this.porsalida++;
        }
        return casilla_final;
    }
    
    int CalcularTirada(int origen, int destino){
        int casilla_final = destino - origen;
        if(casilla_final < 0){
            casilla_final = casilla_final + casillas.size(); 
        }
        return casilla_final;
    }
}


